v1.9: Add new `retrieveVectors` and `_vectors` behaviour (#2876)
meilisearch/documentation: reference/api/documents.mdx

]
```

#### `_vectors`

`_vectors` is a special document attribute containing vector embeddings for AI-powered search. It accepts different types of values.

> `_vectors` may be an array of numbers:

Reviewer: No, `_vectors` is always a map, whose keys are the name of embedders (or future embedders).

The value of each key can be an array or an object.

```json
[
  {
    "id": 452,
    "title": "Female Trouble",
    "overview": "Delinquent high school student Dawn Davenport runs away from home and embarks upon a life of crime.",

>     "_vectors": [0.0, 0.1, 0.003, 0.75]

Reviewer: ```suggestion
    "_vectors": { "default":  [0.0, 0.1, 0.003, 0.75] }
```

>     "_vectors": [0.0, 0.1, 0.003, 0.75]

Reviewer 2: Don't we want to promote this format instead?
`_vectors: { default: [...] } }

@dureuill

Reviewer: Hello @curquiza 

this format is incorrect and will not work as soon as an embedder config has actually been declared.

See my comments below

  },
]
```

> `_vectors` may also be an object with two fields, `embeddings` and `regenerate`:

Reviewer: ```suggestion
Each subobject of `_vectors` may also be an object with up to two fields, `embeddings` and `regenerate`:
```

```json
[
  {
    "id": 452,
    "title": "Female Trouble",
    "overview": "Delinquent high school student Dawn Davenport runs away from home and embarks upon a life of crime.",
    "_vectors": {

>       "embeddings": [],

Reviewer: It would be nicer to have an actual embedding here, even with only 3 floats.

```suggestion
      "embeddings": [0.1, 0.2, 0.3],
```

      "regenerate": true
    }
  }
]
```

> `embeddings` must be an array of numbers.

Reviewer: ```suggestion
If provided, `embeddings` must be an array of numbers, representing a single embedding for that embedder for that document, or an array of arrays of numbers, representing multiple embeddings for that embedder for that document.

If missing or `null`, the embeddings for that document will remain at their current value, except if `regenerate` is `true` and other fields in the document were modified in the same document update.
```

I suggest we introduce `regenerate` first.

Author: Could you expand on this? I find the explanation flows better by saying "embeddings defines the vectors for this document" -> "regenerate regenerates the embeddings your previously defined".

Reviewer: Sure.

- `embeddings` overrides the current embeddings of the documents
- separately from this, `regenerate` describes the behavior when the document is updated.

So, the flow could be:

By default, Meilisearch will generate embeddings when a document is added and each time the document is updated such that the filling the `documentTemplate` with the document's data produces a different text as before the update.

When setting `regenerate` to `false,` this behavior will be disabled, and the embedding will remain the same regardless of the document's updates.

Additionally, if providing the `embeddings` field, the embeddings for that document and embedder will be overridden. 

The various combinations of `regenerate` and `embeddings` can achieve various use cases:

### 1. Providing embeddings to a `userProvided` embedder

Setting `regenerate` to `false` and the `embeddings` to the target value will achieve this use case.

### 2. Importing embeddings that were created externally using a similar embedder once, then letting regeneration occurs when updating documents.

Import the embeddings in `embeddings`, and set `regenerate` to `true` so that future updates regenerate the embeddings.

### 3. Opting-out of embeddings for a single document.

Similar to use case (1), with `regenerate` set to `false`, but with an `embeddings` equal to `[]`

### 4. "Freezing" the current values of embeddings.

Pass only `regenerate` to `false` without specifying `embeddings`. The document will keep the embeddings currently in its database, regardless of future updates.

Author: I'm still digesting this, but a question. Are these two examples equivalent?

```json
{
  "default": [0.1, 0.2, 0.3]
}
```

```json
{
  "default": {
    "embeddings": [0.1, 0.2, 0.3],
    "regenerate" : true
}
```

If they're identical, what is the use case for the simple array format? Wouldn't it be better to remove it and simplify the API, so users don't need to think which format is best for them?

Author: Ah, ok, so by default Meilisearch will not regenerate custom embeddings. That makes sense, thanks.

I think you have found the correct framing for this. The new format (with embeddings and regenerate) is the actual format, whilst the the regular array is simply a shorthand. Idk, perhaps it was obvious to you, but I kept seeing them as two separate things.

> `regenerate` must be a Boolean value. If `regenerate` is `true`, the supplied embeddings will be overwritten the next time Meilisearch automatically generates document embeddings. If `regenerate` is `false`, the document's embeddings will be automatically replaced by an embedder.

Reviewer: ```suggestion
`regenerate` must be a Boolean value. If `regenerate` is `true`, embeddings will be regenerated for that document and embedder when the document is updated. If `regenerate` is `false`, the last value of the embeddings will be kept on document updates.
```

> If `_vectors` is empty or `null`, Meilisearch treats the document as not having any embeddings. This document is then ignored during AI-powered searches.

Reviewer: ```suggestion
If the key for an embedder inside of `_vectors`  is empty or `null`, Meilisearch treats the document as not having any embeddings for that embedder. This document is then ignored during AI-powered searches.
```

  {
    "id": 452,
    "title": "Female Trouble",
    "overview": "Delinquent high school student Dawn Davenport runs away from home and embarks upon a life of crime.",
    "_vectors": {
      "default": [0.003, 0.1, 0.75],
      "ollama": [0.1, 0.2, 0.3]
    }
  }
]
```

Keys of the `_vectors` object may also correspond to an object with two fields, `embeddings` and `regenerate`:

```json
[
  {
    "id": 452,
    "title": "Female Trouble",
    "overview": "Delinquent high school student Dawn Davenport runs away from home and embarks upon a life of crime.",
    "_vectors": {
      "default": {
        "embeddings": [0.1, 0.2, 0.3],
        "regenerate": true
      }
    }
  }
]
```

`embeddings` must be an array of numbers representing a single embedding for that document. It may also be an array of arrays of numbers representing multiple embeddings for that document. `embeddings` is an optional field.

`regenerate` must be a Boolean value. If `regenerate` is `true`, Meilisearch automatically generates embeddings for that document immediately and every time the document is updated. If `regenerate` is `false`, Meilisearch keeps the last value of the embeddings on document updates.

> If `_vectors` is empty or `null`, Meilisearch treats the document as not having any embeddings. This document is then ignored during AI-powered searches.

Reviewer: > If `_vectors` is empty or `null`, Meilisearch treats the document as not having any embeddings.

I don't think that's true. I think this paragraph can probably be removed since we have the contents of the capsule already

<Capsule intent="note">

> If the key for an embedder inside `_vectors` is empty or `null`, Meilisearch treats the document as not having any embeddings for that embedder. This document is then ignored during AI-powered searches.

Reviewer: ```suggestion
If the key for an embedder inside `_vectors` is empty or `null`, Meilisearch treats the document as not having any embeddings for that embedder. This document is then returned last during AI-powered searches.
```

Ah, sorry for not noticing earlier. The current behavior of the semantic search is same as the geodistance and any  sort: documents that don't fulfill the condition (don't have an embedding for that embedder, don't have a geopoint, don't have the sorted field) are returned after any document that does fulfill the condition in an unspecified order.

    "overview": "Delinquent high school student Dawn Davenport runs away from home and embarks upon a life of crime.",
    "_vectors": {
      "default": {
        "embeddings": [0.1, 0.2, 0.3],
        "regenerate": false
      },
      "ollama": {
        "embeddings": [0.4, 0.12, 0.6],
        "regenerate": true
      }
    }
  }
]
```

`embeddings` is an optional field. It must be an array of numbers representing a single embedding for that document. It may also be an array of arrays of numbers representing multiple embeddings for that document. `embeddings` defaults to `null`.

`regenerate` is a mandatory field. It must be a Boolean value. If `regenerate` is `true`, Meilisearch automatically generates embeddings for that document immediately and every time the document is updated. If `regenerate` is `false`, Meilisearch keeps the last value of the embeddings on document updates.

You may also use an array shorthand to add embeddings to a document:

```json
"_vectors": {
  "default": [0.003, 0.1, 0.75]
}
```

Vector embeddings added with the shorthand are not replaced when Meilisearch generates new embeddings. The above example is equivalent to:

```json
"default": {

>   "embeddings": [0.1, 0.2, 0.3],

Reviewer: We should probably use the same vector value here, otherwise it is a bit confusing 🤣 
```suggestion
  "embeddings": [0.003, 0.1, 0.75],
```
